flocker/node/agents/ebs.py

```diff
@@ -1245,6 +1254,11 @@ def detach_volume(self, blockdevice_id):
         if ebs_volume.state != VolumeStates.IN_USE.value:
             raise UnattachedVolume(blockdevice_id)
 
+        if ebs_volume.attachments:
+            attachment_state = ebs_volume.attachments[0]['State']
+            if attachment_state == VOLUME_ATTACHMENT_BUSY:
+                raise DetachFailed(ebs_volume)
```

Comment:
The EC2 API won't return an error if attachment state is busy when you attempt to detach, rather it will report detaching state:

```
{'ResponseMetadata': {'HTTPStatusCode': 200, 'RequestId': 'a52cd151-b8a9-4ee1-b91f-4d355c65acb1'}, u'AttachTime': datetime.datetime(2016, 1, 5, 10, 2, 7, tzinfo=tzutc()), u'InstanceId': 'i-fcc27026', u'VolumeId': 'vol-986c4e7e', u'State': 'detaching', u'Device': '/dev/sdf'}
```

However, the detach will not happen and you will continue to see actual attachment state of e.g.

```
[{u'AttachTime': datetime.datetime(2016, 1, 5, 10, 2, 7, tzinfo=tzutc()), u'InstanceId': 'i-fcc27026', u'VolumeId': 'vol-986c4e7e', u'State': 'busy', u'DeleteOnTermination': False, u'Device': '/dev/sdf'}]
```